### PR TITLE
Only mark projection as ambiguous if GAT substs are constrained

### LINKED ISSUE #93874: #92917 breaks type inference on GATs
Seems to be caused by #92917 @jackh726 
> Not a fix, but a mitigation to prevent a backwards-compatible hazard where we normalize using a predicate only because it's the only one available, but shouldn't.

So this regression seems to be intentional, but I want to make sure because even very basic inference fails now, making any use of GATs very fill up with explicit type annotations. It doesn't look so bad in this example, but with complex types it's bad, and I think there's a more complex example the compiler can't even tell what `_x` is at all, but I'm still working on that.

### Code

```rust
#![feature(generic_associated_types)]

pub trait Build {
    type Output<O>;
    fn build<O>(self, input: O) -> Self::Output<O>;
}

pub struct IdentityBuild;
impl Build for IdentityBuild {
    type Output<O> = O;
    fn build<O>(self, input: O) -> Self::Output<O> {
        input
    }
}

fn a() {
    let _x: u8 = IdentityBuild.build(10);
    // ^type mismatch resolving `<IdentBuild as Build>::Output<i32> == u8`
}

fn b() {
    let _x: Vec<u8> = IdentityBuild.build(Vec::new());
    // ^type annotations needed: cannot satisfy `<IdentBuild as Build>::Output<Vec<_>> == Vec<u8>`
    // cannot satisfy `<IdentBuild as Build>::Output<Vec<_>> == Vec<u8>`
}

pub fn main() {
    a();
    b();
}
```

I expected to see this happen: _compiles without error_

Instead, this happened: Type inference fails, `Output<O> = O` seems to be ignored or not understood by the compiler.

### Version it worked on

It most recently worked on: 
* cc38176793e9e13bb7b70dde4b951d9371017662
* nightly-2022-02-09

### Version with regression

* 9a5a961be97f405e751dd2cf966e1cdb80a612c2
* nightly-2022-02-10

`rustc --version --verbose`:
```
rustc 1.60.0-nightly (e7aca8959 2022-02-09)
binary: rustc
commit-hash: e7aca895980f25f6d2d3c48e10fd04656764d1e4
commit-date: 2022-02-09
host: x86_64-pc-windows-msvc
release: 1.60.0-nightly
LLVM version: 13.0.0
```

### Backtrace
(no crash)

### PR DESCRIPTION
**Pull request by compiler-errors**
A slightly more targeted version of #92917, where we only give up with ambiguity if we infer something about the GATs substs when probing for a projection candidate.

fixes #93874
also note (but like the previous PR, does not fix) #91762

r? @jackh726 
cc @nikomatsakis who reviewed #92917

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 1]
```diff
@@ -1545,7 +1549,24 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                     nested_obligations.into_iter().chain(obligations),
                 )
                 .map_or(false, |res| res.may_apply())
-            })
+            });
+
+        if is_match {
+            let generics = self.tcx().generics_of(obligation.predicate.item_def_id);
+            if !generics.params.is_empty() {
+                // If any of the obligation's predicate substs shallow-resolve to
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:1557]: Should I leave a note here saying that this is still kinda a hack?

**Comment by jackh726** [comment]
This is probably right, but need to take another look (and want Niko's thoughts)

**Comment by MingweiSamuel** [comment]
I tested this patch and I confirmed it fixed #93874 in our project (has decently complex but not pathological use of GATs)

**Review by jackh726 — CHANGES_REQUESTED** [review]

**Comment by jackh726** [comment]
@bors r=jackh726,nikomatsakis

We talked about this during the GATs meeting

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 2]
```diff
@@ -1508,12 +1508,15 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         })
     }
 
+    /// Return Some(true) if the obligation's predicate type applies to the env_predicate, and
+    /// Some(false) if it does not. Returns None in the case that the projection type is a GAT,
+    /// and applying this env_predicate constrains any of the obligation's GAT substitutions.
     pub(super) fn match_projection_projections(
```
Comment (marmeladema) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:1514]: Comment is slightly outdated now that the return type has changed to an enum

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:1514]: Good catch.

**Comment by compiler-errors** [comment]
@jackh726 this might need a new r+

**Comment by marmeladema** [comment]
@compiler-errors I think you updated stdarch in your recent commit but I don't think it was on purpose? Maybe it's just a GitHub trick though

**Comment by compiler-errors** [comment]
> @compiler-errors I think you updated stdarch in your recent commit but I don't think it was on purpose? Maybe it's just a GitHub trick though

No I'm dumb. Fixed it like a second after you commented.

File: compiler/rustc_trait_selection/src/traits/project.rs [hunk 3]
```diff
@@ -1180,7 +1171,7 @@ fn assemble_candidates_from_trait_def<'cx, 'tcx>(
         ProjectionCandidate::TraitDef,
         bounds.iter(),
         true,
-    )
+    );
```
Comment (jackh726) [inline compiler/rustc_trait_selection/src/traits/project.rs:1174]: Is this diff new?

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/project.rs:1174]: function returns () but and the rest have semicolons, but i'll double check

**Pull request merged**